app/models/miq_server/worker_management/heartbeat.rb

```diff
@@ -75,13 +81,23 @@ def post_message_for_workers(class_name = nil, resync_needed = false, sync_messa
 
   # Get the latest heartbeat between the SQL and memory (updated via DRb)
   def validate_heartbeat(w)
+    last_heartbeat = workers_last_heartbeat(w.pid)
+
     if w.last_heartbeat.nil?
-      w.last_heartbeat ||= @workers[w.pid][:last_heartbeat] if @workers.key?(w.pid)
-      w.last_heartbeat ||= Time.now.utc
-    else
-      if @workers.key?(w.pid) && !@workers[w.pid][:last_heartbeat].nil? && @workers[w.pid][:last_heartbeat] > w.last_heartbeat
-        w.update_attributes(:last_heartbeat => @workers[w.pid][:last_heartbeat])
-      end
+      last_heartbeat ||= Time.now.utc
+      w.update_attributes(:last_heartbeat => last_heartbeat)
+    elsif !last_heartbeat.nil? && last_heartbeat > w.last_heartbeat
+      w.update_attributes(:last_heartbeat => last_heartbeat)
+    end
+  end
+
+  private
+
+  def workers_last_heartbeat(pid)
+    return unless @workers_lock
```

Comment:
I don't believe so @Fryguy, I modeled it after the existing pessimistic logic.  I can remove it.